codegen/smithy-dafny-codegen/src/main/java/software/amazon/polymorph/smithygo/localservice/DafnyLocalServiceGenerator.java

```diff
@@ -303,8 +338,29 @@ func NewClient(clientConfig $L) (*$T, error) {
                     symbolProvider.toSymbol(outputShape)
                   )
                 );
+            returnError = "return nil,";
           }
-          returnError = "return nil,";
+        }
+        String validationCheck = "";
+        // inputShape is not a structure only when we have a positional trait
+        // TODO: Figure out how to validate constraint in a structure with positional trait
+        if (
+          !inputShape.hasTrait(UnitTypeTrait.class) &&
+          inputShape.isStructureShape()
+        ) {
+          validationCheck =
+            """
+                err := params.Validate()
+                if err != nil {
+                  opaqueErr := %s.OpaqueError{
```

Comment:
Out of curiosity, is the behavior in Go to return an Opaque Error for invalid input?

In Java, we thrown an `IllegalArguementException`. 
I do not know what .NET or Python do.

You could throw the local service's default exception.
i.e: in the MPL, [`AwsCryptographicMaterialProvidersException`](https://github.com/aws/aws-cryptographic-material-providers-library/blob/89701fe7582f336ab7caddc25c256b3bf2aaf1c1/AwsCryptographicMaterialProviders/dafny/AwsCryptographicMaterialProviders/Model/material-provider.smithy#L83-L87).

That would be marginally better.
Because `Opaque` is intended for unknown errors.
Where as we clearly know that root cause here: bad input.

Reply:
It's a bit more complicated in Go.

Our customer facing API doesn't impose any checks on what errors we can return. But our wrapped shim does. And Dafny Go doesn't use generics so there's no way to know what error we have received in the shim.

There are bunch of work arounds and we could use the base exception but I think there's not much benefit to reap compared to the re-architecting that we might have to do.